### PR TITLE
Implemented FDK

### PR DESCRIPTION
**Pull request by D1rk123**
Added FDK implementation that is using the existing FBP implementation for filtering and backprojecting

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -94,6 +94,7 @@ examples folder.
 ## Authors and contributors
 
 * **Allard Hendriksen** - *Initial work*
+* **Dirk Schut** - *FDK implementation*
```
Comment (ahendriksen) [inline README.md:97]: Great!

File: tests/test_fdk.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+def make_box_phantom():
+    x = torch.ones((32, 32, 32))
+    x[4:28, 4:28, 4:28] = 1.0
+    x[12:22, 12:22, 12:22] = 0.0
+    return x
+
+
+def test_fdk_reconstruction():
+    vg = ts.volume(shape=32, size=1)
+    pg = ts.cone(angles=32, shape=(48, 48), size=(2, 2), src_det_dist=2, src_orig_dist=2)
```
Comment (ahendriksen) [inline tests/test_fdk.py]: Please test with at least 64^3 voxels and 64^2 pixels. I have observed weird behaviour with ASTRA with fewer pixes/voxels.

File: tests/test_fdk.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    A = ts.operator(vg, pg)
+    x = make_box_phantom()
+    y = A(x)
+
+    # very rough test if reconstruction is close to original volume
+    rec = fdk(A, y)
+    assert torch.allclose(rec, x, atol=1, rtol=0)
```
Comment (ahendriksen) [inline tests/test_fdk.py]: Does this test if each reconstructed voxel's intensity differs from the phantom by at most `1`? I would think that `0.5` or `0.1` should also work here?

File: ts_algorithms/fdk.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            weights_mat *= (src_to_obj_n_dist / src_det_n_dist)
+
+        result[:, i, :] = projections[:, i, :] * weights_mat
+
+    return result
+
+
+def fdk(A, y, padded=True, filter=None, reject_acyclic_filter=True, recalculate_weights=False):
+    """Approximately reconstruct volumes in a circular cone beam geometry using
+    the Feldkamp, Davis and Kress(FDK) algorithm
```
Comment (ahendriksen) [inline ts_algorithms/fdk.py]: It is useful to refer to the literature for more details: 

- Feldkamp, L. A., Davis, L. C., & Kress, J. W. (1984). Practical Cone-Beam
  Algorithm. Journal of the Optical Society of America A, 1(6), 612.
  http://dx.doi.org/10.1364/josaa.1.000612

File: ts_algorithms/fdk.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+    If `y` is located on GPU, the entire algorithm is executed on a single GPU.
+
+    If `y` is located in RAM (CPU in PyTorch parlance), then only the
+    foward and backprojection are executed on GPU.
+
+    :param A: `tomosipo.operator`
+    :param y: `torch.tensor`
```
Comment (ahendriksen) [inline ts_algorithms/fdk.py]: I see that I have not done this myself, but I would recommend stating the expected size of `y` as a sinogram stack, i.e., "y has size `(num_vertical_pixels, num_angles, num_horizontal_pixels)`.

File: ts_algorithms/fdk.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    :param filter: bool, is passed to ts_algorithms.fbp
+    :param reject_acyclic_filter: bool, is passed to ts_algorithms.fbp
+    :param recalculate_weights: bool, experimental feature: recalculate FDK weights
+    every angle when using non-circular geometries
+    :returns: reconstruction of a volume
+    :rtype: `torch.tensor`
+
+    """
+    voxel_sizes = np.array(A.volume_geometry.size) / np.array(A.volume_geometry.shape)
+    if (np.ptp(voxel_sizes) > ts.epsilon):
```
Comment (ahendriksen) [inline ts_algorithms/fdk.py]: `np.ptp` is neat! I didn't know this function.

File: ts_algorithms/fdk.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+
+def fdk_weigh_projections(op, projections, recalculate_weights):
+    dev = projections.device
+    u_num_pixels = op.range_shape[2]
+    u_space = (torch.arange(u_num_pixels, dtype=torch.float, device=dev)+0.5-(u_num_pixels/2))
+    v_num_pixels = op.range_shape[0]
+    v_space = (torch.arange(v_num_pixels, dtype=torch.float, device=dev)+0.5-(v_num_pixels/2))
+
+    result = torch.empty(projections.shape, dtype=torch.float, device=dev)
```
Comment (ahendriksen) [inline ts_algorithms/fdk.py]: I would recommend using `projections.new_empty` to match the dtype and device of `projections`

File: ts_algorithms/fdk.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    for i in range(op.range_shape[1]):
+        if recalculate_weights or i == 0:
+            src_to_det = torch.from_numpy(op.range.det_pos[i] - op.range.src_pos[i]).float().to(dev)
+            src_to_obj = torch.from_numpy(-op.range.src_pos[i]).float().to(dev)
```
Comment (ahendriksen) [inline ts_algorithms/fdk.py]: Geometry parameters are 64-bit float. Does converting them to float32 immediately impact the result (I expect not)?

File: ts_algorithms/fdk.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+            u_pos_squared = (u_space * det_u_norm + src_det_u_dist)**2
+            v_pos_squared = (v_space * det_v_norm + src_det_v_dist)**2
+
+            weights_mat = src_det_n_dist / torch.sqrt(
+                  u_pos_squared[None, :].expand([v_num_pixels, u_num_pixels])
+                  + v_pos_squared[:, None].expand([v_num_pixels, u_num_pixels])
+                  + src_det_n_dist**2
+            )
+            # scale detector to origin
+            weights_mat *= (src_to_obj_n_dist / src_det_n_dist)
```
Comment (ahendriksen) [inline ts_algorithms/fdk.py]: Why is this necessary? And does this still work when the center of rotation is not on the origin?

Comment (ahendriksen) [reply ts_algorithms/fdk.py]: I wrote this test, which fails for me: 

``` python


def test_fdk_translation_invariance():
    vg = ts.volume(shape=32, size=1)
    pg = ts.cone(
        angles=32, shape=(48, 48), size=(2, 2), src_det_dist=2, src_orig_dist=2
    ).to_vec()

    # Create sinogram and reconstruct at the origin
    A0 = ts.operator(vg, pg)
    x = make_box_phantom()
    y = A0(x)
    rec0 = fdk(A0, y)
    bp0 = A0.T(y)

    # Move volume and detector far away
    T = ts.translate((1000, 1000, 1000))
    A = ts.operator(T * vg, T * pg)

    rec = fdk(A, y)
    bp = A.T(y)

    # Check that backprojection still matches
    assert torch.allclose(bp, bp0)
    # Check that fdk still matches (fails!)
    assert torch.allclose(rec, rec0)

```

Comment (ahendriksen) [reply ts_algorithms/fdk.py]: Scaling invariance does hold up, so that is good! 

``` python 



def test_fdk_scaling_invariance():
    vg = ts.volume(shape=32, size=1)
    pg = ts.cone(
        angles=32, shape=(48, 48), size=(2, 2), src_det_dist=2, src_orig_dist=2
    ).to_vec()

    # Create sinogram and reconstruct at the origin
    A0 = ts.operator(vg, pg)
    x = make_box_phantom()

    rec0 = fdk(A0, A0(x))
    bp0 = A0.T(A0(x))

    s = 2.0
    S = ts.scale(s)
    A = ts.operator(S * vg, S * pg)

    rec = fdk(A, A(x))
    bp = A.T(A(x))

    # Check that backprojection still matches (up to scaling)
    assert torch.allclose(bp / s**2, bp0)
    # Check that fdk still matches (fails!)
    assert torch.allclose(rec, rec0, atol=0.5, rtol=0.01)
```

**Review by ahendriksen — CHANGES_REQUESTED** [review]
It already looks good! 

Some changes are required though. The most important issue right now is that the FDK reconstruction is not invariant under a rigid transformation of the volume and projection geometry --- for instance when translating both volume and projection geometry in the same direction. 

All in all, the code looks good. I do not really understand fdk_weigh_projections in detail yet. So if you could sprinkle some high-level comments through the function, that would really help me. 

Also, I see that the rest of the code base is not at the same high level of quality that I am holding you to. My apologies: I should improve the rest of the code base..

File: tests/test_fdk.py (hunk 3 above)
Comment (D1rk123) [reply tests/test_fdk.py]: Yes, there turn out to be pretty big errors near edges. I looked at the images and they looked normal. I replaced allclose (which is basically a maximum error metric) with the mean absolute error so I can use a tighter bound.

File: ts_algorithms/fdk.py (hunk 8 above)
Comment (D1rk123) [reply ts_algorithms/fdk.py]: The main reason why I converted everything to 32bit float is so that you could do these computations on the GPU, this is not possible with 64 bit floats. You could of course also calculate the weights mat on the CPU and transfer it to the used device afterwards. I don't expect there to be much of a speed or accuracy difference, but I guess it would be better to err on the side of caution and do the calculations as 64 bits on the cpu. What do you think?

**Comment by D1rk123** [comment]
I'm still not perfectly happy with how the center of rotation has to be passed as an extra argument when using a cone_vec geometry, but I don't think that information is available otherwise. You could in theory fit a circle on all object positions to determine it's center of rotation but that seemed like a worse solution to me.

File: ts_algorithms/fdk.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+def fit_src_rot_center_dist(A):
+    vg = A.domain
+    pg = A.range
+
+    T = ts.from_perspective(
+        pos=pg.det_pos,
+        w=pg.det_v/np.linalg.norm(pg.det_v[0]),
+        v=pg.det_normal/np.linalg.norm(pg.det_normal[0]),
+        u=pg.det_u/np.linalg.norm(pg.det_u[0])
+    )
```
Comment (D1rk123) [inline ts_algorithms/fdk.py:58]: @ahendriksen This part could be made even shorter by adding a Transform.normalize_scale() function that normalizes the w, v and u vectors to 1

**Comment by ahendriksen** [comment]
Hi, 

Thanks for your patience and for investigating how to properly implement FDK. I have taken your commits and reworded some of the commit messages and merged them. 

See: https://github.com/ahendriksen/ts_algorithms/commits/master

I found that the correct center of rotation can be calculated quite robustly. Therefore I have removed the option to specify it manually.
